File: Sources/iTunesKit/Core/iTunesCatalogResponse.swift
//
//  iTunesCatalogResponse.swift
//  iTunesKit
//
//  Created by Aarav Gupta on 23/03/26.
//

import Foundation

// MARK: - Apple Music Catalog API Models
public struct iTunesCatalogResponse: Codable, Sendable {
    public let data: [iTunesCatalogData]
    public let resources: Resources
}

// MARK: - Data
public struct iTunesCatalogData: Codable, Sendable {
    public let id: String
    public let type: String
    public let href: String
}

// MARK: - Resources
public struct Resources: Codable, Sendable {
    public let albums: [String: Album]
    public let songs: [String: Song]
}

// MARK: - Album
public struct Album: Codable, Sendable {
    public let id: String
    public let type: String
    public let href: String
    public let attributes: AlbumAttributes
}

public struct AlbumAttributes: Codable, Sendable {
    public let editorialVideo: EditorialVideo
}

// MARK: - Song
public struct Song: Codable, Sendable {
    public let id: String
    public let type: String
    public let href: String
    public let attributes: SongAttributes
    public let relationships: SongRelationships?
}

public struct SongAttributes: Codable, Sendable {
    public let hasLyrics: Bool
    public let name: String
    public let url: String
}

public struct SongRelationships: Codable, Sendable {
    public let albums: RelationshipAlbums
}

public struct RelationshipAlbums: Codable, Sendable {
    public let href: String
    public let data: [iTunesCatalogData]
}

public struct EditorialVideo: Codable, Sendable {
    public let motionDetailSquare, motionDetailTall, motionSquareVideo1X1, motionTallVideo3X4: MotionVideo

    enum CodingKeys: String, CodingKey {
        case motionDetailSquare, motionDetailTall
        case motionSquareVideo1X1 = "motionSquareVideo1x1"
        case motionTallVideo3X4 = "motionTallVideo3x4"
    }
}

public struct MotionVideo: Codable, Sendable {
    public let previewFrame: PreviewFrame
    public let video: String
}

public struct PreviewFrame: Codable, Sendable {
    public let bgColor: String
    public let hasP3: Bool
    public let height: Int
    public let textColor1, textColor2, textColor3, textColor4: String
    public let url: String
    public let width: Int
}


// This file was generated from JSON Schema using quicktype, do not modify it directly.
// To parse the JSON, add this file to your project and do:
//
//   let iTunesCatalogResponse = try? JSONDecoder().decode(ITunesCatalogResponse.self, from: jsonData)

File: ituneskit/src/main/java/aaravgupta/ituneskit/core/iTunesCatalogResponse.kt
package aaravgupta.ituneskit

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
public data class iTunesCatalogResponse(
    val data: List<iTunesCatalogData>,
    val resources: Resources
)

@Serializable
public data class iTunesCatalogData(
    val id: String,
    val type: String,
    val href: String
)

@Serializable
public data class Resources(
    val albums: Map<String, Album>,
    val songs: Map<String, Song>
)

@Serializable
public data class Album(
    val id: String,
    val type: String,
    val href: String,
    val attributes: AlbumAttributes
)

@Serializable
public data class AlbumAttributes(
    val editorialVideo: EditorialVideo
)

@Serializable
public data class Song(
    val id: String,
    val type: String,
    val href: String,
    val attributes: SongAttributes,
    val relationships: SongRelationships? = null
)

@Serializable
public data class SongAttributes(
    val hasLyrics: Boolean,
    val name: String,
    val url: String
)

@Serializable
public data class SongRelationships(
    val albums: RelationshipAlbums
)

@Serializable
public data class RelationshipAlbums(
    val href: String,
    val data: List<iTunesCatalogData>
)

@Serializable
public data class EditorialVideo(
    val motionDetailSquare: MotionVideo,
    val motionDetailTall: MotionVideo,
    @SerialName("motionSquareVideo1x1")
    val motionSquareVideo1X1: MotionVideo,
    @SerialName("motionTallVideo3x4")
    val motionTallVideo3X4: MotionVideo
)

@Serializable
public data class MotionVideo(
    val previewFrame: PreviewFrame,
    val video: String
)

@Serializable
public data class PreviewFrame(
    val bgColor: String,
    val hasP3: Boolean,
    val height: Int,
    val textColor1: String,
    val textColor2: String,
    val textColor3: String,
    val textColor4: String,
    val url: String,
    val width: Int
)